oauthenticator/generic.py

```diff
@@ -91,11 +97,13 @@ async def authenticate(self, handler, data=None):
 
         headers = {"Accept": "application/json", "User-Agent": "JupyterHub"}
 
-        if self.basic_auth:
+        if self.basic_auth and self.client_auth_method == "client_secret_basic":
```

Comment:
According to https://tools.ietf.org/html/rfc6749#section-2.3.1
If a client is using `basic_auth` it should not be sending credentials in the body, so I think this should be just `if self.basic_auth:`

Alternatively we could deprecate `basic_auth` and replace it with `client_auth_method: 'basic' (default) | 'client_secret_post'`, or something like that?

@consideRatio @minrk any thoughts?

Reply:
I think the `basic_auth` flag should be deprecated along with this, and that we transition to a system based on choosing `client_auth_method` where the allowed options are a subset of the OpenID specifications allowed options. Since they include more than things starting with `client_secret_` i suggest we don't abbreviate this to just `basic` but let it remain `client_secret_basic` to be more transparent and easier to connect to the specification we try to follow.

> If a client is using basic_auth it should not be sending credentials in the body, so I think this should be just if self.basic_auth:

With self.basic_auth deprecated in favor of self.client_auth_method, I figure `self.basic_auth` and `self.client_auth_method == "client_secret_basic"` can be asserted to be the same value in deprecation logic, and that we check only for `self.client_auth_method == "client_secret_basic"`

---

I think the hardest part for me working with this repo are the environment variables. I lack an overview on what environment variables do for libraries used. I'm currently afraid of removing it for example, because perhaps it influence something else, and I'm also afraid of introducing the new `OAUTH2_AUTH_METHOD` not knowing if it has meaning or not for another library.

```
    basic_auth = Bool(
        os.environ.get('OAUTH2_BASIC_AUTH', 'True').lower() in {'true', '1'},
        config=True,
        help="Disable basic authentication for access token request",
    )
```

Reply:
As far as I know several of the environment variables were required because they used to be the only way to set some properties, but @minrk did some work to refactor this so all properties can now be set as traitlets. I don't think environment variables are needed for new properties, and we could perhaps remove the existing env-vars in a future breaking release (but that's a separate discussion).